graylog2-server/src/main/java/org/graylog2/outputs/BatchedMessageFilterOutput.java

```diff
@@ -0,0 +1,235 @@
+        this.daemonScheduler = daemonScheduler;
+
+        this.batchSize = metricRegistry.histogram(name(this.getClass(), "batchSize"));
+        this.bufferFlushes = metricRegistry.meter(name(this.getClass(), "bufferFlushes"));
+        this.bufferFlushFailures = metricRegistry.meter(name(this.getClass(), "bufferFlushFailures"));
+        this.bufferFlushesRequested = metricRegistry.meter(name(this.getClass(), "bufferFlushesRequested"));
+        this.processTime = metricRegistry.timer(name(this.getClass(), "processTime"));
+        this.outputWriteFailures = metricRegistry.meter(name(this.getClass(), "outputWriteFailures"));
+
+        this.buffer = new IndexSetAwareMessageOutputBuffer(outputBatchSize);
+    }
+
+    @Override
+    public void initialize() throws Exception {
+        LOG.debug("Starting buffer flush task to run every {} milliseconds", outputFlushInterval.toMillis());
+        this.flushTask = daemonScheduler.scheduleAtFixedRate(() -> {
+            try {
+                if (buffer.shouldFlush(outputFlushInterval)) {
+                    forceFlush();
+                }
+            } catch (Exception e) {
+                LOG.error("Caught exception while trying to flush outputs", e);
+            }
+        }, outputFlushInterval.toMillis(), outputFlushInterval.toMillis(), TimeUnit.MILLISECONDS);
+    }
+
+    @VisibleForTesting
+    void forceFlush() {
+        LOG.debug("Force-flushing the buffer");
+        bufferFlushesRequested.mark();
+        buffer.flush(this::flush);
+    }
+
+    private void flush(List<FilteredMessage> filteredMessages) {
+        if (filteredMessages.isEmpty()) {
+            if (LOG.isTraceEnabled()) {
+                LOG.trace("Skipping buffer flush with empty buffer");
+            }
+            return;
+        }
+
+        batchSize.update(filteredMessages.size());
+
+        activeFlushThreads.incrementAndGet();
+        if (LOG.isTraceEnabled()) {
+            LOG.trace("Starting flushing {} messages, flush threads active {}",
+                    filteredMessages.size(),
+                    activeFlushThreads.get());
+        }
+
+        try (var ignored = processTime.time()) {
+            for (final var output : outputs.values()) {
+                try {
+                    if (LOG.isTraceEnabled()) {
+                        LOG.trace("Writing {} message(s) to output <{}>", filteredMessages.size(), output);
+                    }
+                    output.writeFiltered(filteredMessages);
+                } catch (Exception e) {
+                    LOG.error("Couldn't write {} message(s) to output <{}>", filteredMessages.size(), output.getClass(), e);
+                    outputWriteFailures.mark();
+                }
+            }
+
+            bufferFlushes.mark();
+            // We only acknowledge messages once all outputs are done writing messages.
+            acknowledger.acknowledge(filteredMessages.stream().map(FilteredMessage::message).toList());
```

Comment:
Could we add an `acknowledge(Stream<Message> messages)` method to the acknowledger?

That method just does a forEach on the list, seems unnecessary to construct an entire list here again.

Reply:
Good point. I will create a separate PR for that. There seems to be some duplication between `AbstractMessageQueueAcknowledger` and `LocalKafkaMessageQueueAcknowledger` as well when it comes to handling lists of messages.

Reply:
Good point. I will create a separate PR for that. There seems to be some duplication between `AbstractMessageQueueAcknowledger` and `LocalKafkaMessageQueueAcknowledger` as well when it comes to handling lists of messages.